charms/kfp-api/src/charm.py

```diff
@@ -286,98 +202,180 @@ def set_pod_spec(self, event):
+
+    def _get_object_storage(self, interfaces):
+        relation_name = "object-storage"
+        return self.validate_sdi_interface(interfaces, relation_name)
+
+    def _get_viz(self, interfaces):
+        relation_name = "kfp-viz"
+        default_viz_data = {"service-name": "unset", "service-port": "1234"}
+        return self.validate_sdi_interface(interfaces, relation_name, default_return=default_viz_data)
+
+    def validate_sdi_interface(self, interfaces: dict, relation_name: str, default_return=None):
+        """Validates data received from SerializedDataInterface, returning the data if valid
+
+        Optionally can return a default_return value when no relation is established
+
+        Raises:
+            CheckFailed(..., Blocked) when no relation established (unless default_return set)
+            CheckFailed(..., Blocked) if interface is not using SDI
+            CheckFailed(..., Blocked) if data in interface fails schema check
+            CheckFailed(..., Waiting) if we have a relation established but no data passed
+
+        Params:
+            interfaces:
+
+        Returns:
+              (dict) interface data
+        """
+        # If nothing is related to this relation, return a default value or raise an error
+        if relation_name not in interfaces or interfaces[relation_name] is None:
+            if default_return is not None:
+                return default_return
+            else:
+                raise CheckFailed(f"Missing required relation for {relation_name}", BlockedStatus)
+
+        relations = interfaces[relation_name]
+        if not isinstance(relations, SerializedDataInterface):
+            raise CheckFailed(f"Unexpected error with {relation_name} relation data - data not as expected", BlockedStatus)
+
+        # Get and validate data from the relation
+        try:
+            # relations is a dict of {(ops.model.Relation, ops.model.Application): data}
+            unpacked_relation_data = relations.get_data()
+        except ValidationError as val_error:
+            # Validation in .get_data() ensures if data is populated, it matches the schema and is
+            # not incomplete
+            self.log.error(val_error)
+            raise CheckFailed(
+                f"Found incomplete/incorrect relation data for {relation_name}.  See logs",
+                BlockedStatus
+            )
+
+        # Check if we have an established relation with no data exchanged
+        if len(unpacked_relation_data) == 0:
+            raise CheckFailed(f"Waiting for {relation_name} relation data", WaitingStatus)
+
+        # Unpack data (we care only about the first element)
+        data_dict = list(unpacked_relation_data.values())[0]
+
+        # Catch if empty data dict is received (JSONSchema ValidationError above does not raise
+        # when this happens)
+        if len(data_dict) == 0:
+            raise CheckFailed(f"Found incomplete/incorrect relation data for {relation_name}.", BlockedStatus)
```

Comment:
This seems more like a bug in SDI than something that we should be handling as an edge case in our charms.

Reply:
Agreed.  I don't want to take it out because I think it can happen, but I'll also file it on SDI to make sure it is prevented